cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -40,63 +48,63 @@
   private Map<String, Map<String, Long>> mapReduceCounters;
   private final Map<Scope, Map<String, List<NodeValue>>> tokenValueMap = new EnumMap<>(Scope.class);
   private String nodeName;
+  private final Set<String> nodeFilterSet;
 
   public BasicWorkflowToken() {
     for (Scope scope : Scope.values()) {
-      tokenValueMap.put(scope, new HashMap<String, List<NodeValue>>());
+      tokenValueMap.put(scope, new ConcurrentHashMap<String, List<NodeValue>>());
     }
+
+    nodeFilterSet = new HashSet<>();
   }
 
-  private BasicWorkflowToken(BasicWorkflowToken other) {
+  private BasicWorkflowToken(BasicWorkflowToken other, String nodeName, Set<String> nodeFilterSet) {
     for (Map.Entry<Scope, Map<String, List<NodeValue>>> entry : other.tokenValueMap.entrySet()) {
-      Map<String, List<NodeValue>> tokenValueMapForScope = new HashMap<>();
+      Map<String, List<NodeValue>> tokenValueMapForScope = new ConcurrentHashMap<>();
       for (Map.Entry<String, List<NodeValue>> valueEntry : entry.getValue().entrySet()) {
-        tokenValueMapForScope.put(valueEntry.getKey(), Lists.newArrayList(valueEntry.getValue()));
+        tokenValueMapForScope.put(valueEntry.getKey(),
+                                  Collections.synchronizedList(Lists.newArrayList(valueEntry.getValue())));
       }
 
       this.tokenValueMap.put(entry.getKey(), tokenValueMapForScope);
     }
 
-    this.nodeName = other.nodeName;
+    this.nodeName = nodeName;
+
+    this.nodeFilterSet = ImmutableSet.copyOf(nodeFilterSet);
 
     if (other.mapReduceCounters != null) {
       this.mapReduceCounters = copyHadoopCounters(other.mapReduceCounters);
     }
   }
 
-  void setCurrentNode(String nodeName) {
-    this.nodeName = nodeName;
-  }
-
   /**
    * Merge the other WorkflowToken passed to the method as a parameter
    * with the WorkflowToken on which the method is invoked.
+   * @param nodeName the name of the current node
    * @param other the other WorkflowToken to be merged
    */
-  void mergeToken(BasicWorkflowToken other) {
+  void mergeToken(final String nodeName, BasicWorkflowToken other) {
     for (Map.Entry<Scope, Map<String, List<NodeValue>>> entry : other.tokenValueMap.entrySet()) {
-      Map<String, List<NodeValue>> thisTokenValueMapForScope = this.tokenValueMap.get(entry.getKey());
 
-      for (Map.Entry<String, List<NodeValue>> otherTokenValueMapForScopeEntry : entry.getValue().entrySet()) {
-        if (!thisTokenValueMapForScope.containsKey(otherTokenValueMapForScopeEntry.getKey())) {
-          thisTokenValueMapForScope.put(otherTokenValueMapForScopeEntry.getKey(), Lists.<NodeValue>newArrayList());
-        }
+      ConcurrentHashMap<String, List<NodeValue>> thisTokenValueMapForScope =
```

Comment:
Why do you cast here? If this class makes explicit use of concurrent features, then the type should be declared as ConcurrentMap, not Map